### ISSUE TITLE
`MissingMethodException` when using new C# language features (`in` parameters, `init`-only setters, etc.) with generic methods, or with members in generic types

### ISSUE THREAD

**Issue by stakx** [devlooped/moq#1148]
**TL;DR:** Upgrade to .NET 6+ (including Preview versions) to get your code working.

We've got quite a few reports lately of `MissingMethodException` happening for generic methods, or methods in generic types, when those methods make use of newer C# language features, e.g. `in` parameters, or `init`-only setters. I'm going to close all of those reports and pin this issue at the top of the issue tracker to make this info easier to find.

Some more background:

The issue is caused by a bug inside System.Reflection.Emit (https://github.com/dotnet/runtime/issues/25958), which causes method signatures to be reproduced incorrectly if the original method signature happens to include a [modreq](https://github.com/stakx/ecma-335/blob/master/docs/ii.7.1.1-modreq-and-modopt.md#ii711-modreq-and-modopt), and if the method is either generic or in a generic type. Lately, the IL metadata encodings for new C# language features increasingly rely on such modreqs (see e.g. [here](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/init#metadata-encoding) for `init`-only setters, or [here](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.2/readonly-ref#metadata-representation-of-in-parameters) for `in` parameters), that's why they're affected by this bug.

The bad news is, because the bug is in the runtime, there's nothing we can do in Moq 4 to solve this problem. A limited workaround inside Castle DynamicProxy (the library underneath Moq) would be theoretically possible (https://github.com/castleproject/Core/issues/430#issuecomment-756309040) but very costly and tricky to implement, so that might not happen.

The good news is, the bug has since been fixed in .NET 6 (https://github.com/dotnet/runtime/pull/40587) so once you upgrade, your code should once again work correctly.

**Comment by KinsonDigital** [comment]
@stakx Thanks for this post!!  It has given me some progress, but I still have some issues after I upgrade the projects to **NET 6.01.00-preview.3**.

I **was having this issue using the method signature below:**
`unsafe void BufferSubData<T>(GLBufferTarget target, nint offset, nuint size, in T data) where T : unmanaged;`

The problem was occurring for me as long as I had the combination of the method being generic as well as the 4th parameter `data` is an `in` parameter.

---

**Current New Issue:**
Once I updated to **NET 6** that issue was resolved now, my **Moq** verify is failing and it should not be.

**Here is my verify method call:**
`this.mockGL.Verify(m => m.BufferSubData(GLBufferTarget.ArrayBuffer, It.IsAny<nint>(), It.IsAny<nuint>(), It.IsAny<QuadData>()));`

**And this is what I am verifying:**
`this.gl.BufferSubData(GLBufferTarget.ArrayBuffer, (nint)offset, this.totalQuadSizeInBytes, this.quadData);`

This should be passing but it is failing.  I am not testing for actual data because I am using `It.IsAny<T>()` for all of the parameters but the first one.

Is this possibly still related to the issue in NET 5 runtime?

Thanks!!

**Comment by stakx** [comment]
@KinsonDigital could you please post this as a separate issue? Make sure to include minimal but complete repro code, along with the error message you received. Thanks!

P.S.: Due to a bug in Castle DynamicProxy, `nuint` could be problematic (see https://github.com/castleproject/Core/issues/546). If that's indeed the culprit, replacing it with `nint` might make the problem go away.

**Comment by KinsonDigital** [comment]
Absolutely!! I will also give the native int pointer change a shot as well.  

Thanks!!

**Comment by KinsonDigital** [comment]
Just a quick update for transparency and to possibly help anybody out.  I changed the data type from `nuint` to `nint` and the problem still exists.

**Comment by stakx** [comment]
.NET 6 has been out for fairly long and is now an LTS release, so this reminder to update to .NET 6 is perhaps no longer necessary. Closing.
